src/Elastic.Apm/ServerInfo/ServerInfo.cs

```diff
@@ -0,0 +1,78 @@
+// Licensed to Elasticsearch B.V under
+// one or more agreements.
+// Elasticsearch B.V licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information
+
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Elastic.Apm.Config;
+using Elastic.Apm.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Elastic.Apm.ServerInfo
+{
+	/// <summary>
+	/// A "real" <see cref="IServerInfo" /> implementation.
+	/// The <see cref="GetServerInfoAsync" /> method must be called once, which sets the <see cref="Version" /> property.
+	/// </summary>
+	internal class ServerInfo : IServerInfo
+	{
+		private readonly IConfigSnapshot _configSnapshot;
+		private readonly IApmLogger _logger;
+
+		public ServerInfo(IConfigSnapshot configSnapshot, IApmLogger logger) => (_configSnapshot, _logger) = (configSnapshot, logger);
+
+		private volatile bool _initialized;
+		public bool ServerVersionQueried => _initialized;
+		public Version Version { get; private set; }
+
+		public async Task GetServerInfoAsync()
```

Comment:
Should the fetching of server info (behaviour) be separate from the data?

Reply:
I separated the fetch logic - one related problem I found is that if an API keys or Secret token is set, then it should be also used for the Server Info API, otherwise the APM server won't respond.